src/aes_circuit.rs

```diff
@@ -2,17 +2,24 @@ use crate::helpers::traits::ToAnyhow;
 use anyhow::{ensure, Result};
 use ark_r1cs_std::{
     prelude::{AllocVar, Boolean},
+    select::CondSelectGadget,
     R1CSVar, ToBitsGadget,
 };
 use collect_slice::CollectSlice;
-use simpleworks::gadgets::{ConstraintF, UInt32Gadget, UInt8Gadget};
+use simpleworks::{
+    gadgets::{ConstraintF, UInt32Gadget, UInt8Gadget},
+    marlin::ConstraintSystemRef,
+};
 
 /// This function returns the derived keys from the secret key.
 /// Because AES 128 consists of 11 rounds, the result are 11 128-bit keys,
 /// which we represent as 4 32-bit words, so we compute 44 32-bit elements
 /// W_0, W_1, ..., W_43. The first four constitute the first round key, the
 /// second four the second one, and so on.
-pub fn derive_keys(secret_key: &[UInt8Gadget]) -> Result<Vec<Vec<UInt8Gadget>>> {
+pub fn derive_keys(
```

Comment:
I see that this function returns a `Result` only because be are turning every `Option` into one. Idk if it is worth doing but if this is the case we could avoid this and do something like we did in `shift_rows`

Reply:
But maybe it is better to return a `Result` to indicate that the function failed.

Reply:
There are calls inside `derive_keys` to functions like
```
UInt32Gadget::new_constant(
            constraint_system.clone(),
            u32::from_be_bytes([0x01, 0x00, 0x00, 0x00]),
        )?
```
that return a `Result` though

Reply:
You're right!